rust/projects/parallelism/project.md

```diff
@@ -0,0 +1,447 @@
+`kvs-client`. As with previous projects, add enough definitions that the
+test suite builds.
+
+
+## Part 1: Multithreading
+
+Your first try at introducing parallelism is going to be the simplest: spawning
+a new thread per incoming connection, and responding to the request on that
+connection, then letting the thread exit. What performance benefits will
+distributing work across threads provide? How do you expect latency will be
+affected? What about throughput?
+
+The first step is to write a `ThreadPool` implementation for this naive
+approach. Call it `NaiveThreadPool` (it's not _really_ even a thread _pool_
+since the threads are not reused, but it needs to conform to our trait for later
+comparisons).
+
+We aren't focusing on a more sophisticated implementation now because simply
+integrating this solution into our existing design is going to take some effort.
+So note that the `ThreadPool::new` constructor takes a `threads` argument
+specifying the number of threads in the pool. In this implementation it will go
+unused.
+
+_Go ahead and implement this version of `ThreadPool` now_, then we'll
+integrate it into the new `KvStore`.
+
+**Test cases to complete**:
+
+  - `naive_thread_pool_*`
+
+
+## Part 2: Creating a shared `KvEngine`
+
+Before we can integrate the `NaiveThreadPool` into `KvServer` we have to make
+the `KvEngine` trait and the `KvStore` implementation (for now you can ignore
+the `SledKvsEngine` from the previous project, but you can optionally
+re-implement it as an extension to this project).
+
+Recall from the project spec that, this time, our `KvsEngine` takes `self` as
+`&self` instead of `&mut self` as previously, and it also implements `Clone`.
+More concretely, it looks like
+
+```rust
+pub trait KvsEngine: Clone {
+    fn set(&self, key: String, value: String) -> Result<()>;
+
+    fn get(&self, key: String) -> Result<Option<String>>;
+
+    fn remove(&self, key: String) -> Result<()>;
+}
+```
+
+This gives us a lot of clues about the implementation strategy we're pursuing.
+First, think about why the engine needs to implement `Clone` when we have a
+multi-threaded implementation. Instead of being coy, this once I'll just tell
+you that each thread is going to need access to the engine. Now think about why
+that makes us use `&self` instead of `&mut self`. What do you know about shared
+mutable state? By the end of this project be sure you understand the
+implications here &mdash; _this is what Rust is all about_.
+
+In this model, `KvsEngine` behaves like a _handle_ to another object, and
+because that object is shared between threads, it probably needs to live on the
+[heap], and because that shared state can't be mutable it needs to be protected by
+some synchronization primitive.
+
+[heap]: https://stackoverflow.com/questions/79923/what-and-where-are-the-stack-and-heap
+
+So, _hoist the date inside your implementation of `KvEngine`, `KvsStore` onto
```

Comment:
I don't understand "hoist the date"... Is it a slang?

Reply:
it should have said `data`, but "hoist" is slang. It means "pull up" or "pull out": "pull up your data from the stack to the heap". I'll change it.

Reply:
more precisely, "hoist" isn't _slang_ but using it in this way is jargony.